This Jupyter notebook takes the polynomials computed in `collect_polynomials.ipynb` to determine the irreducibility, up to complex conjugation, of realization spaces of simple $(3,11)$\-matroids with the 3 lines property.

# Univariate polynomials

We first look at the 3658 cases of simple $(3,11)$\-matroids with the 3 lines property where there exists a maximal circuit such that the ideal of the realization sp  
ace is univariate. These ideals are all generated by a single univariate polynomial. To determine the realization space up to complex conjugation, we look at the number of complex roots, up to complex conjugation, of a polynomial that generates the ideal of the realization space.  



In [1]:
from sage.all import *

# load univariate polynomials
univariate_path = 'univariate_polys.txt'
with open(univariate_path, 'r') as file:
    uni_poly_str = file.read()
    
uni_poly_str = uni_poly_str.replace('x1', 'x')
uni_poly_str = uni_poly_str.replace('x2', 'x')

uni_poly_str = uni_poly_str.split('\n')[:-1]

In [2]:
R = PolynomialRing(QQ, 'x')
n = len(uni_poly_str)
comp_conj = []  # tracks indices that have complex conjugate realization spaces
reducible = []  # tracks indices that have reducible realization spaces.
irreducible = []  # tracks indices that that have irreducible realization spaces.
for i in range(n):
    poly = R(uni_poly_str[i])
    poly_factored = poly.factor()
    n = len(poly_factored)  # number of factors over Q, without counting multiplicity
    
    if n > 1:  
        reducible.append(i)
    else:  # polynomial irreducible over Q
        poly_red = poly_factored[0][0]
        if poly_red.degree() < 2:
            irreducible.append(i)
        elif poly_red.degree() == 2:
            d = discriminant(poly)
            if d < 0:
                comp_conj.append(i)
            else:  # discriminant = 0 can't happen for a factor irreducible over Q
                reducible.append(i)
        else:
            reducible.append(i)  

In [6]:
with open("uni_comp_conj.txt", 'w') as file:
    for i in comp_conj:
        file.write(str(i))
        file.write("\n")
        
with open("uni_irreducible.txt", 'w') as file:
    for i in irreducible:
        file.write(str(i))
        file.write("\n")
        
with open("uni_reducible.txt", 'w') as file:
    for i in reducible:
        file.write(str(i))
        file.write("\n")

In [3]:
len(comp_conj)

808

In [4]:
len(irreducible)

0

In [5]:
len(reducible)

2850

Now we summarize the 3658 cases of simple $(3,11)$\-matroids with the 3 lines property where there exists a maximal circuit such that the ideal of the realization space is univariate. There are

- 808 cases where the realization space is two irreducible complex conjugate components, making these realization spaces irreducible up to complex conjugation,
- 0 cases where the realization space is a single irreducible component, and
- 2850 cases where the realization space is at least two irreducible components up to complex conjugation, making these realization spaces reducible up to complex conjugation.   



# Multivariate polynomials

Now we take a look at the 5638 cases of simple $(3,11)$\-matroids with the 3 lines property where there exists a maximal circuit such that the ideal of the realization space is multivariate and principal and that there doesn't a maximal circuit so that the ideal of the realization space is univariate. To determine the irreducibility of these realization spaces, we look at a multivariate polynomial that generates the ideal of the realization space. We factor these polynomials over $\mathbb{C}$ and identify the complex conjugate factors.   



In [66]:
# load multivariate polynomials
multivariate_path = 'multivariate_polys.txt'
with open(multivariate_path, 'r') as file:
     multi_poly_str = file.read()
    
multi_poly_str = multi_poly_str.split('\n')[:-1]

In [67]:
R.<x1,x2,x3,x4> = QQbar[]  # the polynomials have at most 4 variables

In [111]:
from itertools import combinations

n = len(multi_poly_str)
multi_comp_conj = []  # tracks indices that have complex conjugate realization spaces
multi_reducible = []  # tracks indices that have reducible and not connected realization spaces.
multi_reducible_connected = []  # tracks indices that have reducible but connected realization spaces.
multi_irreducible = []  # tracks indices that that have irreducible realization spaces.
for i in range(n):
    poly = R(multi_poly_str[i])
    vs = poly.variables()
    A = QQbar[vs]
    poly = A(multi_poly_str[i])
    
    poly_factored = poly.factor()
    n_comp_factors = len(poly_factored)  # number of factors over Qbar, without counting multiplicity
    
    if n_comp_factors > 1:
        poly_real = poly.change_ring(AA)
        poly_real_factored = poly_real.factor()
        n_real_factors = len(poly_real_factored)  # number of factors over algebraic reals, without counting multiplicity
        
        if n_real_factors == 1:
            multi_comp_conj.append(i)
        else:
            intersection = []
            for pair in combinations(range(n_real_factors), 2):
                p1 = poly_real_factored[pair[0]][0]
                p2 = poly_real_factored[pair[1]][0]
                I = A.ideal(p1, p2)
                intersection.extend(I.variety())

            if len(intersection) > 0:
                multi_reducible_connected.append(i)
            else:
                multi_reducible.append(i)
    else:
        multi_irreducible.append(i)  

In [113]:
with open("multi_comp_conj.txt", 'w') as file:
    for i in multi_comp_conj:
        file.write(str(i))
        file.write("\n")
        
with open("multi_irreducible.txt", 'w') as file:
    for i in multi_irreducible:
        file.write(str(i))
        file.write("\n")
        
with open("multi_reducible.txt", 'w') as file:
    for i in multi_reducible:
        file.write(str(i))
        file.write("\n")
        
with open("multi_reducible_connected.txt", 'w') as file:
    for i in multi_reducible_connected:
        file.write(str(i))
        file.write("\n")

In [112]:
print(len(multi_comp_conj), len(multi_irreducible), len(multi_reducible), len(multi_reducible_connected))

62 5525 9 42


Out of the 5638 cases of simple $(3,11)$\-matroids with the 3 lines property where there exists a maximal circuit such that the ideal of the realization space is multivariate and principal \(and that there doesn't a maximal circuit so that the ideal of the realization space is univariate\), there are 

- 62 cases where the realization space is two irreducible complex conjugate components, making the realization space irreducible up to complex conjugation,
- 5525 cases where the realization space is one irreducible component, making the realization space irreducible up to complex conjugation, and
- 51 cases where the realization is at least two irreducible components up to complex conjugation, making the realization space reducible up to complex conjugation. 

For the 51 cases where the realization space is at least two irreducible components up to complex conjugation, there are

- 9 cases where the irreducible components don't share any points pairwise and
- 42 cases where at least two of the irreducible components share a point.   

The 42 cases where at least two of the irreducible components can be given further analysis as to whether the common points correspond to degenerate realizations.
